Цель работы: изучение связи между признаками двумерного набора данных, визуализация данных.

Общее задание:
1. Загрузить датасет с помощью библиотеки pandas. Оценить его через info и describe. 
Выполнить предварительную обработку данных, если это необходимо.

2. Построить точечную диаграмму (матрицу диаграмм рассеяния) для всех признаков.
Выполнить анализ полученной диаграммы, отвечая на вопрос показывает ли она в
среднем определенную зависимость между переменными. Изучите параметры и
опишите взаимосвязи. Если параметров слишком много – может потребоваться
создать несколько графиков. Минимум один график (диаграмму рассеивания)
сделать по категориям (Например, зависимость зарплаты и возраста по каждой
должности. На таком графике различные должности должны быть показаны разными
цветами).

3. Постройте гистограммы для каждого числового признака, подберите оптимальное
количество bins, сделайте выводы по полученными гистограммам.

4. Исследовать взаимосвязь между переменными с помощью оценки коэффициента
корреляции и ковариации. Построить heatmap (тепловую карту корреляции).
Выполнить интерпретацию результатов корреляции и ковариации, отвечая на вопросы 
о наличии (отсутствии) линейной взаимосвязи между переменными. Понимать, что
такое корреляция и ковариация.

5. Постройте графики по заданию в варианте. Выполните интерпретацию, что вы видите
на этих графиках.

6. Выполните минимум один любой график типа hexagonal binning plot. Сделайте
выводы.

7. Выполните минимум один график типа boxplot для любого столбца. Сделайте выводы.

8.  Добавьте категорию по любому числовому столбцу (например, уровень зарплаты -
высокий, низкий, средний). Сделайте boxplot по этому числовому столбцу по каждой
новой категории (на boxplot будет box-ы для средней, низкой и высокой зарплаты).

9. Выполните ещё минимум 2 графика boxplot по другим категориям (например,
зарплата по полу, зарплата по должности и т.п. - в результате несколько box-ов для
каждой категории на ном графике). Используйте для построения графиков разные
библиотеки (минимум 2). Сделайте вывод.

10. Сделайте общие выводы по работе. В том числе с качественной интерпретацией.

Индвидидуальное задание - вариант 7:

Набор данных salary2.csv

Содержит информацию о зарплатах.
1. Год выплаты заработной платы (целое число)
2. Тип работы (PT - Part-time, FT - Full-time, FL - Freelance)
3. Должность
4. Зарплата за год (целое число)
5. Зарплата в долларах (целое число)
6. Страна проживания
7. Страна главного офиса
8. Среднее кол-во людей в компании (S - менее 50 сотрудников (малая), M от 50 до 250 сотрудников (средняя), L - более 250 сотрудников (крупная))
9. Общий опыт работы
10. Опыт на последней работе

Задание 1: Использовать seaborn. По группировке - work_year и количество
компаний по различному типу занятости (employment_type) построить диаграмму
следующего вида:

<img src="task1.png" alt="" height="300"/>

Задание 2: Использовать pandas и plot. По сводной таблице (pivot_table) -
отобразить среднюю зарплату в usd по локациям (company_location). Сделать
линию зелёной, шириной 4 и отобразить маркеры в виде *.

<img src="task2.png" alt="" height="300"/>

Задание 3: Использовать matplotlib. Построить круговую диаграмму, которая
отображает процент количества записей по каждому году.

<img src="task3.png" alt="" height="300"/>


# 1. Предварительный анализ данных и обработка

## 1.1. Чтение набора данных

В начале работы были подключены необходимые библиотеки для анализа и визуализации данных (pandas, seaborn, matplotlib). Затем был считан CSV-файл salary2.csv в датафрейм pandas для дальнейшей обработки и анализа

In [342]:
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt

# считывание файл в датафрейм.
# используется sep для определения символа разделителя в датасете,
# нужно чтобы датасет загрузился в правильном формате.
df = pd.read_csv('salary2.csv', sep=';')

## 1.2 Первичная оценка данных

Для ознакомления с набором данных, были использованы методы head, info и describe.

In [343]:
print(df.head())
print()
print(df.info())
print()
print(df.describe())

   work_year employment_type             job_title      salary  salary_in_usd  \
0     2020.0              FT        Data SCIENTIST     70000.0        79833.0   
1     2020.0              FT  Product Data Analyst     20000.0        20000.0   
2     2020.0              FT          Data Analyst     72000.0        72000.0   
3     2020.0              FT        Data Scientist  11000000.0        35735.0   
4     2020.0              FT        Data Scientist     45000.0        51321.0   

  employee_residence company_location company_size  exp_all  exp  
0                 DE               DE            L        9    4  
1                 HN               HN            S        2    2  
2                 US               US            L       13    3  
3                 HU               HU            L       60    6  
4                 FR               FR            S        8    4  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401 entries, 0 to 400
Data columns (total 10 columns):
 #   

По выводу программы можно сделать следующие замечания:
1. Тип колонки work_year вещественный, хотя должен быть целочисленным по заданию.
2. Стоит проверить наличие аномалий в колонке employment_type, там могу содержаться неявные дубликаты.
3. Аналогично стоит проверить и колонку job_title, даже в выводе head видно дубликат (Data SCIENTIST - Data Scientist)
4. Колонки salary и salary_in_usd должны быть целочисленными.
5. В колонке salary присутствуют 3 пропуска данных. 
6. Стоит проверить налаичие аномалий в колонке employee_residence, company_location и company_size.
7. Стоит проверить аномалии в связке колонок exp_all и exp. Например слишком большие или маленькие значения, а также не соотвествие между совокупным и последним опытом.

## 1.3 Обработка данных

Сначала из данных были устранены пропуски. Так как пропусков мало, всего 3, то было принято решение удалить такие строчки, а не заменять их значениями.

In [344]:
# удаление строк с пропущенными значениями при помощи dropna
df = df.dropna(subset=['salary'])

# восстановление индексации при помощи reset_index.
# дополнительный столбец index не создается при помощи аргумента drop=True.
df = df.reset_index(drop=True)

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           398 non-null    float64
 1   employment_type     398 non-null    object 
 2   job_title           398 non-null    object 
 3   salary              398 non-null    float64
 4   salary_in_usd       398 non-null    float64
 5   employee_residence  398 non-null    object 
 6   company_location    398 non-null    object 
 7   company_size        398 non-null    object 
 8   exp_all             398 non-null    int64  
 9   exp                 398 non-null    int64  
dtypes: float64(3), int64(2), object(5)
memory usage: 31.2+ KB
None


Затем были исправлены типы данных, при помощи метода astype.

In [345]:
df['work_year'] = df['work_year'].astype('int64')
df['salary'] = df['salary'].astype('int64')
df['salary_in_usd'] = df['salary_in_usd'].astype('int64')

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           398 non-null    int64 
 1   employment_type     398 non-null    object
 2   job_title           398 non-null    object
 3   salary              398 non-null    int64 
 4   salary_in_usd       398 non-null    int64 
 5   employee_residence  398 non-null    object
 6   company_location    398 non-null    object
 7   company_size        398 non-null    object
 8   exp_all             398 non-null    int64 
 9   exp                 398 non-null    int64 
dtypes: int64(5), object(5)
memory usage: 31.2+ KB
None


После были устранены явные дубликаты, а для устранения неявных дубликатов были выведены уникальные значения в рассматриваемых столбцах.

In [346]:
print('количество явных дубликатов')
print(df.duplicated().sum())
print()

# удаление дубликатов (если они есть)
df = df.drop_duplicates()
df = df.reset_index(drop=True)

print('уникальные значения в столбце employment_type')
print(df['employment_type'].unique())
print()

print('уникальные значения в столбце job_title')
print(df['job_title'].unique())
print()

print('уникальные значения в столбце employee_residence')
print(df['employee_residence'].unique())
print()

print('уникальные значения в столбце company_location')
print(df['company_location'].unique())
print()

print('уникальные значения в столбце company_size')
print(df['company_size'].unique())
print()

количество явных дубликатов
0

уникальные значения в столбце employment_type
['FT' 'PT' 'FL']

уникальные значения в столбце job_title
['Data SCIENTIST' 'Product Data Analyst' 'Data Analyst' 'Data Scientist'
 'Data Engineer' 'Machine Learning Manager' 'Data Analytics Engineer'
 'Data Science Engineer' 'Machine Learning Developer'
 'Data Analytics Manager' 'Head of Data Science'
 'Head of Machine Learning' 'NLP Engineer' 'Data Analytics Lead'
 'DataScientist' 'Data AnalyticsManager']

уникальные значения в столбце employee_residence
['DE' 'HN' 'US' 'HU' 'FR' 'IN' 'PK' 'JP' 'GR' 'MX' 'CA' 'AT' 'NG' 'PH'
 'GB' 'ES' 'IT' 'PL' 'BG' 'NL' 'IQ' 'UA' 'SG' 'RU' 'MT' 'CL' 'RO' 'IR'
 'VN' 'BR' 'HK' 'TR' 'RS' 'AR' 'DZ' 'AU' 'CH']

уникальные значения в столбце company_location
['DE' 'HN' 'US' 'HU' 'FR' 'IN' 'PK' 'JP' 'GR' 'MX' 'CA' 'AT' 'NG' 'GB'
 'ES' 'IT' 'LU' 'PL' 'NL' 'IQ' 'UA' 'IL' 'RU' 'MT' 'CL' 'IR' 'BR' 'VN'
 'TR' 'DZ' 'MY' 'AU' 'CH']

уникальные значения в столбце company_size
['L' 'S' 'M'

В выводе программы неявные дубликаты были обнаружены в столбце job_title: Data SCIENTIST - Data Scientist - DataScientist, Data AnalyticsManager - Data Analytics Manager. Дополинтельно неявный дубликаты присутствуют и в столбце company_size - Large и L.

Дубликаты были устранены при помощи метода replace.

In [347]:
# regex=True используется для удобного перечисления значений для замены при помощи прямого слэша '|'
df['job_title'] = df['job_title'].replace('Data SCIENTIST|DataScientist', 'Data Scientist', regex=True)
df['job_title'] = df['job_title'].replace('Data AnalyticsManager', 'Data Analytics Manager')
df['company_size'] = df['company_size'].replace('Large', 'L')

print('уникальные значения в столбце job_title после обработки')
print(df['job_title'].unique())
print()

print('уникальные значения в столбце company_size после обработки')
print(df['company_size'].unique())

уникальные значения в столбце job_title после обработки
['Data Scientist' 'Product Data Analyst' 'Data Analyst' 'Data Engineer'
 'Machine Learning Manager' 'Data Analytics Engineer'
 'Data Science Engineer' 'Machine Learning Developer'
 'Data Analytics Manager' 'Head of Data Science'
 'Head of Machine Learning' 'NLP Engineer' 'Data Analytics Lead']

уникальные значения в столбце company_size после обработки
['L' 'S' 'M']


Последней частью обработки будет исправления аномалий в колонках exp и exp_all. Сначала будут произведены расчеты, а затем выбрана тактика исправлений (удаление/замена).

In [348]:
print('противоречащие связки exp > exp_all:')
print(df[df['exp'] > df['exp_all']][['exp', 'exp_all']])
print()

print('количество отрицательных значений exp:')
print(df[df['exp'] < 0].shape[0])
print()

print('количество отрицательных значений exp_all:')
print(df[df['exp_all'] < 0].shape[0])
print()

print('максимальные exp с их количеством')
# получаем 5 максимальных exp, убираем дубликаты, приводим к python списку.
# далее фильтруем датафрейм по максимальным exp, группируем по ним же, считаем количество строк по каждому максимальному exp.
top5exp = df['exp'].nlargest(5).drop_duplicates().to_list()
top5exp_count = df[df['exp'].isin(top5exp)].groupby('exp').size().sort_values().reset_index(name='count')
print(top5exp_count)
print()

print('максимальные exp_all:')
top5exp_all = df['exp_all'].nlargest(5).drop_duplicates().to_list()
top5exp_all_count = df[df['exp_all'].isin(top5exp_all)].groupby('exp_all').size().sort_values().reset_index(name='count')
print(top5exp_all_count)
print()

противоречащие связки exp > exp_all:
     exp  exp_all
8      4        1
11     8        4
17     2        1
27     7        5
41     6        3
61     6        4
65     7        0
68     9        7
78     7        2
83     7        4
91     6        3
92     1        0
101    7        3
116    8        0
134    4        3
139   10        7
251    8        4
267    6        3
272    8        6
320    7        5
324    3        2

количество отрицательных значений exp:
1

количество отрицательных значений exp_all:
0

максимальные exp с их количеством
   exp  count
0   10      1
1    9      6

максимальные exp_all:
   exp_all  count
0       56      1
1       58      1
2       60      1
3       72      1
4       81      1



In [349]:
# находим все строчки где exp > exp_all.
# с помощью loc заменяем exp_all в этих строчках на exp.
exp_less_all = df['exp'] > df['exp_all']
df.loc[exp_less_all, 'exp_all'] = df['exp']

# удаляем отрицательные значения exp
df = df[df['exp'] >= 0]
df = df.reset_index(drop=True)

print('количество противоречащих связок exp > exp_all:')
print(df[df['exp'] > df['exp_all']][['exp', 'exp_all']].shape[0])
print()

print('количество отрицательных значений exp:')
print(df[df['exp'] < 0].shape[0])
print()

print('количество записей:')
print(df.shape[0])

количество противоречащих связок exp > exp_all:
0

количество отрицательных значений exp:
0

количество записей:
397


В исходных данных было не так много противоречащих связок exp и exp_all, поэтому было принято решение заменить exp_all на exp в случае когда exp > exp_all, хотя данные можно было и удалить.

Максимальные значение exp находятся в пределах нормы, а вот exp_all имеет довольное большие макисмальные значения, причем каждое из которых встречается по 1 разу - 81, 72, 60, 58, 56. Было принято решение не удалять такие данные, допустим что человек может иметь стаж около 80 лет за всюсвою жизнь.

Отрицательные значения exp были удалены.

Итого после обработки осталось 397 значений, то есть удалено всего 4 от исходных 401.

# 2. Точечная диаграмма